exporter/kafkaexporter/kafka_exporter.go

```diff
@@ -117,122 +105,150 @@ func (e *kafkaMetricsProducer) metricsDataPusher(ctx context.Context, md pmetric
 	return nil
 }
 
-func (e *kafkaMetricsProducer) Close(context.Context) error {
-	if e.producer == nil {
-		return nil
+func newTracesExporter(config Config, set exporter.Settings) *kafkaExporter[ptrace.Traces] {
+	// Jaeger encodings do their own partitioning, so disable trace ID
+	// partitioning when they are configured.
+	switch config.Encoding {
+	case "jaeger_proto", "jaeger_json":
+		config.PartitionTracesByID = false
+	}
+	return &kafkaExporter[ptrace.Traces]{
+		cfg:    config,
+		logger: set.Logger,
+		newMessager: func(host component.Host) (kafkaMessager[ptrace.Traces], error) {
+			marshaler, err := getTracesMarshaler(config.Encoding, host)
+			if err != nil {
+				return nil, err
+			}
+			return &kafkaTracesMessager{
+				config:    config,
+				marshaler: marshaler,
+			}, nil
+		},
 	}
-	return e.producer.Close()
 }
 
-func (e *kafkaMetricsProducer) start(ctx context.Context, host component.Host) error {
-	// extensions take precedence over internal encodings
-	if marshaler, errExt := loadEncodingExtension[pmetric.Marshaler](
-		host,
-		e.cfg.Encoding,
-	); errExt == nil {
-		e.marshaler = &metricsEncodingMarshaler{
-			marshaler: *marshaler,
-			encoding:  e.cfg.Encoding,
-		}
-	}
-	if marshaler, errInt := createMetricMarshaler(e.cfg); e.marshaler == nil && errInt == nil {
-		e.marshaler = marshaler
-	}
-	if e.marshaler == nil {
-		return errUnrecognizedEncoding
-	}
-	producer, err := newSaramaProducer(ctx, e.cfg)
-	if err != nil {
-		return err
-	}
-	e.producer = producer
-	return nil
+type kafkaTracesMessager struct {
+	config    Config
+	marshaler marshaler.TracesMarshaler
 }
 
-// kafkaLogsProducer uses sarama to produce logs messages to kafka
-type kafkaLogsProducer struct {
-	cfg       Config
-	producer  sarama.SyncProducer
-	marshaler LogsMarshaler
-	logger    *zap.Logger
+func (e *kafkaTracesMessager) marshalData(td ptrace.Traces) ([]marshaler.Message, error) {
+	return e.marshaler.MarshalTraces(td)
 }
 
-func (e *kafkaLogsProducer) logsDataPusher(ctx context.Context, ld plog.Logs) error {
-	messages, err := e.marshaler.Marshal(ld, getTopic(ctx, &e.cfg, ld.ResourceLogs()))
-	if err != nil {
-		return consumererror.NewPermanent(err)
-	}
-	err = e.producer.SendMessages(messages)
-	if err != nil {
-		var prodErr sarama.ProducerErrors
-		if errors.As(err, &prodErr) {
-			if len(prodErr) > 0 {
-				return kafkaErrors{len(prodErr), prodErr[0].Err.Error()}
+func (e *kafkaTracesMessager) getTopic(ctx context.Context, td ptrace.Traces) string {
+	return getTopic(ctx, &e.config, td.ResourceSpans())
+}
+
+func (e *kafkaTracesMessager) partitionData(td ptrace.Traces) iter.Seq2[[]byte, ptrace.Traces] {
+	return func(yield func([]byte, ptrace.Traces) bool) {
+		if !e.config.PartitionTracesByID {
+			yield(nil, td)
+			return
+		}
+		for _, td := range batchpersignal.SplitTraces(td) {
+			key := []byte(traceutil.TraceIDToHexOrEmptyString(
+				td.ResourceSpans().At(0).ScopeSpans().At(0).Spans().At(0).TraceID(),
```

Comment:
Indexing without checking is hazardous, is this something that could be checked before accessed?

Reply:
Hmm actually I don't know that there's anything to be done: SplitTraces guarantees that each result contains exactly one trace.

Reply:
Added a comment about this. Let me know if you have suggestions to improve.